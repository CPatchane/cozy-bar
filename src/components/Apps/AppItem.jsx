/* global __TARGET__ */

import React from 'react'
import { appShape } from 'proptypes/index'
import { checkApp, startApp } from 'cozy-device-helper'
import expiringMemoize from 'lib/expiringMemoize'
import AppIcon from 'cozy-ui/react/AppIcon'
import stack from 'lib/stack'

const NATIVE_APP_INFOS = {
  drive: {
    appId: 'io.cozy.drive.mobile',
    uri: 'cozydrive://'
  },
  banks: {
    appId: 'io.cozy.banks.mobile',
    uri: 'cozybanks://'
  }
}

const expirationDelay = 10 * 1000
const memoizedCheckApp = expiringMemoize(checkApp, expirationDelay, x => x.appId)

export class AppItem extends React.Component {
  state = {
    isAppAvailable: null
  }

  constructor () {
    super()
    this.openNativeApp = this.openNativeApp.bind(this)
  }

  componentDidMount () {
    if (__TARGET__ === 'mobile') {
      this.checkAppAvailability()
    }
  }

  async checkAppAvailability () {
    const { slug } = this.props.app
    const appInfo = NATIVE_APP_INFOS[slug]
    if (appInfo) {
      const isAppAvailable = Boolean(await memoizedCheckApp(appInfo))
      this.setState({ isAppAvailable })
    }
  }

  openNativeApp (ev) {
    if (ev) {
      ev.preventDefault()
    }
    const appInfos = NATIVE_APP_INFOS[this.props.app.slug]
    startApp(appInfos)
  }

  render () {
    const { app, icon } = this.props
    const dataIcon = app.icon ? `icon-${app.slug}` : ''
    const label = (app.namePrefix ? (app.namePrefix + ' ') : '') + app.name

    let href = app.href
    let onClick = null

    if (!onClick && this.state.isAppAvailable) {
      onClick = this.openNativeApp
    }

    if (onClick) {
      href = '#'
    }

    if (app.isCurrentApp) { // disabled for current app
      onClick = null
      href = null
    }

    return (
      <li className={`coz-nav-apps-item${app.isCurrentApp ? ' --current' : ''}`}>
        <a role='menuitem' href={href} data-icon={dataIcon} title={label} onClick={onClick}>
          {icon
            ? <img
              src={icon}
              className='coz-nav-apps-item-icon'
              alt=''
            />
            : <AppIcon
              app={app}
              className='coz-nav-apps-item-icon'
              fetchIcon={stack.get.icon}
              key={app.slug}
            />
          }
          <p className='coz-label'>{label}</p>
        </a>
      </li>
    )
  }
}

AppItem.propTypes = {
  app: appShape.isRequired
}

export default AppItem
