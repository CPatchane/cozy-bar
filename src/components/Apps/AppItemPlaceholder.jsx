import React from 'react'

export const AppItemPlaceholder = () => (
  <li className="coz-nav-apps-item">
    <span role="menuitem" disabled>
      <div className="coz-nav-apps-item-icon coz-loading-placeholder" />
      <p className="coz-label coz-loading-placeholder" />
    </span>
  </li>
)

export default AppItemPlaceholder
