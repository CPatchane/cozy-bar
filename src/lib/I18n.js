/**
 * preact plugin that provides an I18n helper using a Higher Order Component.
 */

'use strict'

import React, { Component } from 'react'
import Polyglot from 'node-polyglot'

export const DEFAULT_LANG = 'en'

export let _polyglot

const initTranslation = (lang, dictRequire, context, defaultLang = DEFAULT_LANG) => {
  _polyglot = new Polyglot({
    phrases: dictRequire(defaultLang),
    locale: defaultLang
  })

  // Load global locales
  if (lang && lang !== defaultLang) {
    try {
      const dict = dictRequire(lang)
      _polyglot.extend(dict)
      _polyglot.locale(lang)
    } catch (e) {
      console.warn(`The dict phrases for "${lang}" can't be loaded`)
    }
  }

  // Load context locales
  if (context) {
    const dict = dictRequire(lang, context)
    _polyglot.extend(dict)
  }

  return _polyglot
}

// Provider root component
export class I18n extends Component {
  constructor (props) {
    super(props)
    this.init(this.props)
  }

  init (props) {
    const { lang, dictRequire, context, defaultLang } = props

    this.translation = initTranslation(lang, dictRequire, context, defaultLang)
  }

  getChildContext () {
    return {
      t: this.translation.t.bind(this.translation)
    }
  }

  componentWillReceiveProps (newProps) {
    if (newProps.locale !== this.props.locale) {
      this.init(newProps)
    }
  }

  render () {
    return (this.props.children && this.props.children[0]) || null
  }
}

I18n.propTypes = {
  lang: React.PropTypes.string.isRequired,      // current language.
  dictRequire: React.PropTypes.func.isRequired, // A callback to load locales.
  context: React.PropTypes.string,              // current context.
  defaultLang: React.PropTypes.string           // default language. By default is 'en'
}

I18n.childContextTypes = {
  t: React.PropTypes.func
}

// higher order decorator for components that need `t`
export const translate = () => {
  return (WrappedComponent) => {
    const _translate = (props, context) => (
      <WrappedComponent {...props} t={context.t} />
    )
    return _translate
  }
}
