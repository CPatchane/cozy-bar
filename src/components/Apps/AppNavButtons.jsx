import React, { Component } from 'react'
import { connect } from 'react-redux'

import { getHomeApp } from 'lib/reducers'

import Icon from 'cozy-ui/react/Icon'
import { translate } from 'cozy-ui/react/I18n'
import homeIcon from 'assets/icons/icon-cozy-home.svg'

class AppNavButton extends Component {
  render () {
    const { homeApp, busy, onClick, appName, appNamePrefix, iconPath, opened, t } = this.props

    const isHomeApp = homeApp && homeApp.isCurrentApp

    return (
      <div className={`coz-nav-apps-btns${isHomeApp ? ' --currentHome' : ''}`}>
        <a href={homeApp && homeApp.href} className='coz-nav-apps-btns-home'>
          <img src={homeIcon} />
        </a>
        {!isHomeApp && <span className='coz-nav-apps-btns-sep' />}
        <button
          type='button'
          onClick={onClick}
          className='coz-nav-apps-btns-main'
          aria-controls='coz-nav-pop--apps' aria-busy={busy}
          data-tutorial='apps'
        >
          {!isHomeApp && <img className='coz-bar-hide-sm' src={iconPath} width='28' alt='' />}
          <span className='coz-nav-app-name'>
            {isHomeApp
              ? t('menu.home_title')
              : appNamePrefix ? `${appNamePrefix} ${appName}` : appName
            }
          </span>
          {<Icon icon={opened ? 'top' : 'bottom'} color='#95999d' size='12' />}
        </button>
      </div>
    )
  }
}

const mapStateToProps = state => ({
  homeApp: getHomeApp(state)
})

const mapDispatchToProps = dispatch => ({})

export default connect(mapStateToProps, mapDispatchToProps)(translate()(AppNavButton))
