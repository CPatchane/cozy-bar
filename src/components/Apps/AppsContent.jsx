import React, { Component } from 'react'
import { connect } from 'react-redux'

import { translate } from 'cozy-ui/react/I18n'
import withBreakpoints from 'cozy-ui/react/helpers/withBreakpoints'
import { getApps, fetchApps, getHomeApp } from 'lib/reducers'

import AppItem from 'components/Apps/AppItem'
import cozyIcon from 'assets/icons/16/icon-cozy-16.svg'
import homeIcon from 'assets/icons/icon-cozy-home.svg'

class AppsContent extends Component {
  constructor (props, context) {
    super(props, context)
    this.props.fetchApps()
  }

  render () {
    const { t, apps, homeApp, breakpoints } = this.props
    const { isMobile } = breakpoints
    const isHomeApp = homeApp && homeApp.isCurrentApp
    const homeAppWithIcon = Object.assign({}, homeApp, {
      icon: {
        cached: true,
        src: homeIcon
      }
    })

    if (!apps || !apps.length) {
      return <p className='coz-nav--error coz-nav-group'>{t('no_apps')}</p>
    }

    return (
      <div className='coz-nav-pop-content'>
        <ul className='coz-nav-group'>
          {isMobile && homeApp && (
            <AppItem app={homeAppWithIcon} />
          )}
          {apps.map(app => <AppItem app={app} />)}
        </ul>
        {homeApp && !isMobile && !isHomeApp && (
          <a role='menuitem' href={homeApp.href} className='coz-apps-home-btn'>
            <img src={cozyIcon} />
            {t('menu.home')}
          </a>
        )}
      </div>
    )
  }
}

const mapStateToProps = state => ({
  apps: getApps(state),
  homeApp: getHomeApp(state)
})

const mapDispatchToProps = dispatch => ({
  fetchApps: () => dispatch(fetchApps())
})

export default connect(mapStateToProps, mapDispatchToProps)(translate()(withBreakpoints()(AppsContent)))
