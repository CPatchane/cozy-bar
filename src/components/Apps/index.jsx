import React, { Component } from 'react'

import AppsContent from 'components/Apps/AppsContent'
import AppNavButtons from 'components/Apps/AppNavButtons'

class Nav extends Component {
  constructor(props) {
    super(props)
    this.state = {
      busy: false,
      opened: false
    }
    this.toggleMenu = this.toggleMenu.bind(this)
  }

  componentDidMount() {
    document.body.addEventListener('click', this.onClickOutside)
  }

  componentWillUnmount() {
    document.body.removeEventListener('click', this.onClickOutside)
  }

  onClickOutside = event => {
    if (this.state.busy || this.state.opened) {
      // if it's not a cozy-bar nav popup, close the opened popup
      if (!this.rootRef.contains(event.target)) {
        this.setState({ busy: false, opened: false })
      }
      event.stopPropagation()
    }
  }

  toggleMenu = () => {
    this.setState({ opened: !this.state.opened })
  }

  // data-tutorial attribute allows to be targeted in an application tutorial
  render() {
    const {
      replaceTitleOnMobile,
      appName,
      appNamePrefix,
      appSlug,
      iconPath
    } = this.props
    const { busy, opened } = this.state
    return (
      <nav
        className={`coz-nav coz-nav-apps${
          replaceTitleOnMobile ? ' coz-bar-hide-sm' : ''
        }`}
        ref={ref => {
          this.rootRef = ref
        }}
      >
        <AppNavButtons
          appName={appName}
          appNamePrefix={appNamePrefix}
          appSlug={appSlug}
          iconPath={iconPath}
          busy={busy}
          onClick={this.toggleMenu}
          opened={opened}
        />
        <div
          className="coz-nav-pop coz-nav-pop--apps"
          id="coz-nav-pop--apps"
          aria-hidden={!opened}
        >
          <AppsContent />
        </div>
      </nav>
    )
  }
}

export default Nav
