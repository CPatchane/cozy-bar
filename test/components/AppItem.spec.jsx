import React from 'react'
import { AppItem } from 'components/Apps/AppItem'
import { shallow } from 'enzyme'
import { tMock } from '../jestLib/I18n'
import {
  isMobileApp,
  isMobile,
  openDeeplinkOrRedirect
} from 'cozy-device-helper'

jest.useFakeTimers()
jest.mock('lib/stack', () => ({
  get: {
    iconProps: () => {
      return global.__TARGET__ === 'mobile'
        ? { fetchIcon: jest.fn }
        : {
            // we mustn't give the protocol here
            domain: 'cozy.tools',
            secure: true
          }
    }
  }
}))

jest.mock('cozy-device-helper', () => ({
  ...require.requireActual('cozy-device-helper'),
  isMobileApp: jest.fn(),
  isMobile: jest.fn(),
  openDeeplinkOrRedirect: jest.fn()
}))
describe('app icon', () => {
  let spyConsoleError, openNativeSpy

  beforeEach(() => {
    global.__TARGET__ = 'browser'
    spyConsoleError = jest.spyOn(console, 'error')
    spyConsoleError.mockImplementation(message => {
      if (message.lastIndexOf('Warning: Failed prop type:') === 0) {
        throw new Error(message)
      }
    })
    openNativeSpy = jest.spyOn(AppItem.prototype, 'openNativeApp')
    isMobileApp.mockReturnValue(false)
    isMobile.mockReturnValue(false)
  })

  afterEach(() => {
    spyConsoleError.mockRestore()
    openNativeSpy.mockRestore()
  })

  it('should render correctly', () => {
    const app = {
      slug: 'cozy-drive',
      name: 'Drive'
    }
    const root = shallow(<AppItem t={tMock} app={app} />)
    expect(root.getElement()).toMatchSnapshot()
  })

  it('should render correctly with target mobile and providing fetchIcon to AppIcon', () => {
    global.__TARGET__ = 'mobile'
    const app = {
      slug: 'cozy-drive',
      name: 'Drive'
    }
    const root = shallow(<AppItem t={tMock} app={app} />)
    expect(root.getElement()).toMatchSnapshot()
  })

  it('should change the onClick handler if native app is available', () => {
    isMobileApp.mockReturnValue(true)
    const app = {
      slug: 'cozy-drive',
      name: 'Drive'
    }
    const root = shallow(<AppItem t={tMock} app={app} />)
    root.find('a').simulate('click')
    jest.runAllTimers()
    expect(openNativeSpy).not.toHaveBeenCalled()
    root.setState({ isMobileAppAvailable: true })
    root.find('a').simulate('click')
    jest.runAllTimers()
    expect(openNativeSpy).toHaveBeenCalled()
  })

  it('should change the onClick handler to use onAppSwitch if current app is mobile', () => {
    const app = {
      slug: 'cozy-drive',
      name: 'Drive'
    }
    const appSwitchMock = jest.fn()
    const root = shallow(
      <AppItem t={tMock} app={app} onAppSwitch={appSwitchMock} />
    )
    root.find('a').simulate('click')
    jest.runAllTimers()
    expect(appSwitchMock).not.toHaveBeenCalled()
    isMobileApp.mockReturnValue(true)
    root.setState({ isMobileAppAvailable: true })
    root.find('a').simulate('click')
    jest.runAllTimers()
    expect(appSwitchMock).toHaveBeenCalled()
  })

  it('should not change the onClick handler if current app is web and target app is web too', () => {
    const app = {
      slug: 'cozy-store',
      name: 'Store'
    }
    const appSwitchMock = jest.fn()
    const root = shallow(
      <AppItem t={tMock} app={app} onAppSwitch={appSwitchMock} />
    )
    root.find('a').simulate('click')
    jest.runAllTimers()
    expect(appSwitchMock).not.toHaveBeenCalled()
  })

  it('should call deeplink if the currentApp is web mobile and the other app is native', () => {
    isMobile.mockReturnValue(true)
    const app = {
      slug: 'drive',
      name: 'Drive'
    }
    const appSwitchMock = jest.fn()
    const root = shallow(
      <AppItem t={tMock} app={app} onAppSwitch={appSwitchMock} />
    )
    root.find('a').simulate('click', { preventDefault: () => {} })
    jest.runAllTimers()
    expect(openDeeplinkOrRedirect).toHaveBeenCalled()
  })
})
