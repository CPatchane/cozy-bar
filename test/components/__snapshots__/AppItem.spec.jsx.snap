// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app icon should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AppItem
    app={
      Object {
        "name": "Drive",
        "slug": "cozy-drive",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <a
        data-icon=""
        onClick={null}
        role="menuitem"
        title="Drive"
      >
        <AppIcon
          app={
            Object {
              "name": "Drive",
              "slug": "cozy-drive",
            }
          }
          className="coz-nav-apps-item-icon"
          fetchIcon={[Function]}
        />
        <p
          className="coz-label"
        >
          Drive
        </p>
      </a>,
      "className": "coz-nav-apps-item",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <AppIcon
            app={
              Object {
                "name": "Drive",
                "slug": "cozy-drive",
              }
            }
            className="coz-nav-apps-item-icon"
            fetchIcon={[Function]}
          />,
          <p
            className="coz-label"
          >
            Drive
          </p>,
        ],
        "data-icon": "",
        "href": undefined,
        "onClick": null,
        "role": "menuitem",
        "title": "Drive",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "cozy-drive",
          "nodeType": "class",
          "props": Object {
            "app": Object {
              "name": "Drive",
              "slug": "cozy-drive",
            },
            "className": "coz-nav-apps-item-icon",
            "fetchIcon": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Drive",
            "className": "coz-label",
          },
          "ref": null,
          "rendered": "Drive",
          "type": "p",
        },
      ],
      "type": "a",
    },
    "type": "li",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <a
          data-icon=""
          onClick={null}
          role="menuitem"
          title="Drive"
        >
          <AppIcon
            app={
              Object {
                "name": "Drive",
                "slug": "cozy-drive",
              }
            }
            className="coz-nav-apps-item-icon"
            fetchIcon={[Function]}
          />
          <p
            className="coz-label"
          >
            Drive
          </p>
        </a>,
        "className": "coz-nav-apps-item",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <AppIcon
              app={
                Object {
                  "name": "Drive",
                  "slug": "cozy-drive",
                }
              }
              className="coz-nav-apps-item-icon"
              fetchIcon={[Function]}
            />,
            <p
              className="coz-label"
            >
              Drive
            </p>,
          ],
          "data-icon": "",
          "href": undefined,
          "onClick": null,
          "role": "menuitem",
          "title": "Drive",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "cozy-drive",
            "nodeType": "class",
            "props": Object {
              "app": Object {
                "name": "Drive",
                "slug": "cozy-drive",
              },
              "className": "coz-nav-apps-item-icon",
              "fetchIcon": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Drive",
              "className": "coz-label",
            },
            "ref": null,
            "rendered": "Drive",
            "type": "p",
          },
        ],
        "type": "a",
      },
      "type": "li",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
